### DISCUSSION TITLE
WinUI Community Call September 2024 Q&A Follow-up

### DISCUSSION THREAD

**Discussion by duncanmacmichael** [microsoft/microsoft-ui-xaml#9995] [General]
Hi everyone, since we received a lot of questions on the September 2024 community call discussion post and in the call's live comments that we didn't have time to answer, we'd like to follow up here and make sure we address all the ones we couldn't get to during the call. We appreciate your patience and understanding and want to thank you as always for your great feedback and questions. 

> NOTE: If there is a question you asked but don't see here, we either can't answer it, it was a duplicate of another question, or it was already answered on the call itself.

## Questions from the September 2024 community call discussion post

These questions were asked on the [community call discussion post](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946) but we'll consolidate them here along with the other questions in an effort to make the answers clearer and not get lost in individual comment chains.

### Q: Can your team work with the Windows team to enable WinUI3 apps to be easily selectable as Kiosk apps in Windows Settings? It's very easy to do that with my current UWP app. I can't move on to using WinUI3 because of this limitation.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10541229)

**A:** The WinAppSDK is a vehicle for us to ship features from across the company and we depend on each team's priority and schedule to deliver features like this. We are in communication with the team about this need, but they are currently unable to work on this. We will update you on GitHub as we know more.

### Q: WinUI3 has so many memory leaks. Are there any plans to address them at all? As an example #7282. Currently, WinUI never releases memory for second window and the same happens for page navigation.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10548990)

**A:** We agree this needs some attention and plan on investigating this in an upcoming release, although we don't have details for that at the moment.

### Q: Could you share any plans for improving performance with multiple tabs, drag-and-drop functionality, or memory usage? We would appreciate any insights or a sneak peek into what's coming.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10549408)

**A:** Thank you for the question, but we don't have anything to share about plans for these right now.

### Q: Could you guys add input validation?

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10550553)

**A:** Thanks for the feedback! Unfortunately we're unable to prioritize this but we encourage you to keep upvoting the issue as we do take upvotes into account when planning our feature work (as you've seen with the Designer decision).

### Q: Any news for splash screen?

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10558340)

**A:** Splash screen remains on our radar but our current priority is performance. We'd like most apps to launch fast enough so they don't need a splash screen.

### Q: When do you plan to improve and add features to Window? When do you plan to provide methods for attaching UI framework content to an AppWindow?

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10558390)

**A:** As we mentioned on the call, we are planning on adding convenience APIs for XAML Window in the 1.7 time frame.

### Q: "Please priortize quality, tooling support and performance over new features. WinUI should become the choice of UI framework on Windows. This should the goal of the team. I have learned for work item planning for features when I graduated to work. The current status of WinUI requires dedicated bug fixing in planning, with much more effort than new features. With solid framework and vigorous ecosystem, the users and community should be able to implement more features themselves. The unreplacable job for the team is to make the framework solid."

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10561122)

**A:** Windows App SDK 1.6 contains a number of quality and performance improvements (notably, AOT). We are also investing in tooling support. This will continue while we strive to balance with feature development.

### Q: Any plan for cross platform support like Xbox, Hololens etc? Have to say quite amount of users are coming from Xbox revealed by my analytic dashboard. People love using apps on console.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10562961)

**A:** This isn't an area that we're focused on today.

### Q: I believe the big question is what general strategy Microsoft has planned for WinUI ...

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10564527)

**A:** As we communicated at //Build, Microsoft is committed to WinUI and we have shared a view on how Microsoft plans on delivering Windows experiences ranging from Web to Native. We focused our messaging around native apps at //Build 2024 and shared a number of internal apps using the technology, who are part of a growing list. MAUI for Windows sits on top of the Windows App SDK and ReactNative for Windows will also be soon. We've added resources to deliver on community issues such as Designer and improving VS templates, and the recent announcement about preview UWP support for .NET 9 took folks from across the company to deliver. Please be assured that while you might not always see it on the surface, we're putting the efforts in.

### Q: Is there a performance boost even in non-AOT scenarios in WinAppSDK 1.6?

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10565198)

**A:** Can you clarify what scenarios you're referring to? C#? C++? 

### Q: When will we be able to properly use SVGs instead of font icons? See #1494 for all the issues SvgImageSource, PathIconSource, ... have. Another issue is changing the color of SVGs (e.g. on hover, pressed, dark/light) #507.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10571431)

**A:** As with Input Validation, we can't prioritize this right now but we do encourage you to continue upvoting these issues as the number of upvotes does affect our feature planning priorities.

### Q: I would like to confirm the progress of the inkscanvas control, will it arrive in the next version, I remember asking last year. As well as when the input validation will arrive, it's really important. Improving the stability of the framework and continuing to improve responsiveness is a priority. Delays in UI operations always give the impression of poor performance and can affect the user experience bringing a negative impression of the WinUI application.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10576028)

**A:** We're hoping to have a version of InkCanvas available in 1.7. Input Validation is not currently in plan for 1.7.

### Q: Some questions to be answered by the WinUI team \[list of questions including regarding windowing, ProtectedCursor, and more\]

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10590537)

**A:** We can answer some questions here, and for the ones that we don't answer, we don't have plans for them.

1. Window size-limiting properties such as MinHeight, MinWidth, etc. are on our radar for the convenience APIs that we mentioned are in our plans for 1.7.
2. We would like to add a public Cursor property but have not been able to prioritize that. In the meantime, we would like to keep ProtectedCursor protected.
3. We don't have plans to make AppWindow open source.
4. We don't have plans for right-click for InputNonClientPointerSource.
5. For your question about SystemBackdropConfiguration, please see [this comment](https://github.com/microsoft/microsoft-ui-xaml/issues/9915#issuecomment-2305850353).
6. XAML 2009 support is not in our plans but remains on our radar.
7. For more info about UWP and .NET 9, please see our recent blog post [here](https://aka.ms/preview-uwp-support-for-dotnet9).

### Q: I love WinUI - great work. But can you please co-ordinate releases with the other dependancy tooling. For example, the WinUI VS template took nearly a year to go from WinUI release to the VS team; and currently we have the SDK 1.6 release that needs additional components as a workaround because other teams haven't released their complimentary bits yet - the .Net SDK. It makes for a confusing release that seems buggy with temporary workarounds - not a great experience.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10590958)

**A:** Visual Studio 17.12 will include project templates that automatically reference the latest published version of the WinAppSDK and will not need to be refreshed. Periodic updates to the Windows SDK projection might require a temporary WindowsSdkPackageVersion override.

### Q: As a framework, WinUI should avoid exposing unnecessary details to user. After migration from UWP and WinUI, dialogs like file selection and ContentDialog are requiring hwnd or XamlRoot. I understand the technical change because there's no a single AppView bound to the UI thread, but the framework shouldn't expose this to user for simple cases. Just getting the current active window for current thread is sufficient for most cases.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10597657)

**A:** We believe needing to set a XamlRoot is fully necessary for apps to handle, and it's also consistent with Win32 apps needing to specify a parent window. We will continue to work on improving error messages and code analyzers to help identify issues and fixes.

### Q: Microsoft must really demonstrate real dedication to Windows App SDK and Win UI...

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10599627)

**A:** As we mentioned at //Build 2024, we're encouraging apps to use WebView2 if it's right for them, so in a way these teams are eating our own dogfood as the community recommends we do. We do understand that you're expressing pain with this and we appreciate your honest feedback. We'll keep this in mind as we have future conversations with these teams.

### Q: When and in which version will the TitleBar return? Can we see it in the next v1.6.x service update?

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10611111)

**A:** We're planning on releasing the final version of the TitleBar control in 1.7.

### Q: Do they have any further details on how they will be approaching the fix of the visual designer for Visual Studio.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10614371)

**A:** We plan on making progress on Designer during 1.7.

### Q: What is the purpose of creating such a discussion when none of the questions are going to be answered? You wasted our time. Worst call ever. What is the purpose of WinUI 3 development? When you revived both WPF and now UWP?

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10617187)

**A:** While we planned this community call to be full of great content, we unfortunately ran out of time for Q&A as some of the presentations took longer than expected. Balancing content and Q&A was definitely an opportunity on this call that we will work to address for the next one. However, after every community call we do make sure to address all of the questions that we can in follow-up posts like this one, since we value your feedback and questions and want to make sure the community is heard.

We recommend both WinUI and WPF as frameworks for creating native Windows desktop apps, as we showed at //Build 2024. Which framework you choose will depend on your app's needs. With UWP support for .NET 9, we did not revive UWP nor do we recommend creating new apps with it. Instead, it will help you incrementally upgrade your UWP app to WinUI 3 and help you remove dependencies on .NET Native. We are not investing in UWP.

### Q: Omg omg omg I had to re-read the title 10 times... This is the best news for the last ~5 years as a #windows dev and user #uwp #notdead

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10617419)

**A:** We love your enthusiasm for UWP support for .NET 9!

### Q: Microsoft.WindowsAppRuntime.Bootstrap.dll should be included in Windows App SDK runtime

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10620271)

**A:** Microsoft.WindowsAppRuntime.Boostrap.dll must be located within the app, as it initializes dynamic dependencies for the app to find the rest of the runtime in the framework package directory. Placing it in the framework package is a catch-22.

### Q: Is it possible to create CsWinRT authored components (e.g. background task) with PublishAot, while keeping the main app non-AOT?

**A:** Yes.

### Q: how can I compile WinUI 3 project using WindowsAppSDK to Binaries which would result with an exe file that runs on different machines? for now I only managed to make it work for me

**A:** We recommend using WindowsAppSDKSelfContained + PublishSingleFile.

### Q: I'm trying to compile a WinUI 3 app using WindowsAppSDK to binaries, to share with others at the company. The exe file is only working for me, but not for other people. Does anyone know why?  (seems like more a tech support question)

**A:** This is usually caused by a missing dependency. If the app is packaged, a dependent framework package may be missing. Installing the app MSIX is not sufficient, so we recommend zipping up the VS-generated publish folder (with the Install.ps1 script, certificate, and dependencies). If the app is unpackaged, the CRT may need to be installed. Another factor may be whether the app is a debug or release build - for example, the debug UCRT must be manually redistributed.

### Q: Will XAML 2009 be supported in a near future release? XAML 2009 supports TypeArguments, making generic usercontrols MyControl<T> possible.

**A:** This isn't currently in our plans but remains on our radar.

### Q: Since not all libraries/technologies are (yet) compatible with AOT (such as EF, SignalR), will/is it possible to call non-native-AOT compiled libaries from native-AOT compiled code.

**A:** The AOT team recommends that all libraries in an app support AOT. Otherwise, AOT compilation warnings will almost certainly indicate a runtime error.

### Q: When will the new tab tear-out from WASDK 1.6 be implemented in File Explorer and Notepad?

**A:** Unfortunately, we can't announce anything on behalf of apps.

### Q: Is there a plan to add vscode support for WinUI apps in dotnet templates like Uno platform do?

**A:** This isn't currently in our plans.

### Q: \[Regarding the TitleBar control from 1.6-experimental \] Most of that feedback was provided pretty much immediately after the first preview, but still didn't make it into 1.6. Why have previews if you can't reach in time then?

**A:** TitleBar shipped in experimental builds which are subject to change and make no promises of their contents relative to the stable build. We didn't think the control was ready to be made public in 1.6, so it continues to be in experimental until we're ready to ship. We did receive great feedback about it and actually made a change to our process to make spec reviews public because of TitleBar - here's the process if you'd like to learn more: [WinUI Platform Public Spec Review Process](https://github.com/microsoft/microsoft-ui-xaml/blob/main/specs/public-api-review-process.md). The next step will be to submit the TitleBar spec for public review, so keep an eye out for it!

### Q: When you say all classes, you mean like all classes in the project??

**A:** If you're referring to marking classes as partial for AOT, then yes, that is correct per the 1.6 [release notes](https://learn.microsoft.com/windows/apps/windows-app-sdk/stable-channel#partial-classes).

### Q: What I would really love to see is a focus on fixing low-hanging usability issues that haven't been fixed for years.

**A:** We invite you to continue upvoting these issues, as we do take upvotes into account when planning (as indicated by the Designer topic).

### Q: Any plans to modernize the FileOpen and SaveFile dialogs, as was done with the print dialog?

**A:** These aren't in our current plan.

### Q: C#9 features are still not supported, like init properties and required properties. Init and required properties breaking XamlTypeInfo.g.cs generation has been a huge thorn in our side. It means that things like shared DTOs are relegated to C#8

**A:** We have several open issues trackig this (issues [#8723](https://github.com/microsoft/microsoft-ui-xaml/issues/8723), [#5315](https://github.com/microsoft/microsoft-ui-xaml/issues/5315), and [#2519](https://github.com/microsoft/microsoft-ui-xaml/issues/2519)) and encourage you to upvote those issues. If there's something else specific you don't see an issue for, please file a new one.

### Q: Is there any doc on how to config vscode to debug WinUI app?

**A:** We don't have specific advice here yet, but we know this is a community request per GitHub issue [#7331](https://github.com/microsoft/microsoft-ui-xaml/issues/7331) and encourage you to upvote, comment, and discuss this topic there. 

### Q: MinWidth and MaxWidth, MinHeight and MaxHeight???

**A:** These are among the set of APIs we're looking at for XAML Window. See our [1.7 plans post](https://github.com/microsoft/WindowsAppSDK/discussions/4710) for more info.

**Comment by nikolayvpavlov** [comment]
I am utterly disappointed by these answers.  These two:

> MAUI for Windows sits on top of the Windows App SDK and ReactNative for Windows will also be soon.

and

> As we mentioned at //Build 2024, we're encouraging apps to use WebView2 if it's right for them, so in a way these teams are eating our own dogfood as the community recommends we do.

define the sad purpose of Windows App Sdk.  **WASDK is not intended to be the new framework for modern apps.  It is merely an engine to drive WebView2, MAUI and ReactNative.**  Microsoft are clearly not motivated to commit to native apps, WinUI and Fluent design.  There is no leading force, no direction when it comes to UI, and Microsoft Design are powerless to steer the teams into one direction.  Key apps will still be nothing but a web app, masked as native through WebView2, and will all look poor and inconsistent.  Considering the mess the new Outlook is, this is utterly uncomprehensible.

Now, pray tell me, what would I ever want to use WinUI in anything but an island in a WPF app?

**Reply by pjmlp** [reply]
Agree, somehow  WebView2 keeps being mentioned as if it is a WinUI feature, when it only provides a hole to display Web content.

From that point of view, apparently MSHTML ActiveX was Win32 programming.

And no feedback on C++ tooling questions, oh well that ship has sailed apparently.

**Reply by nikolayvpavlov** [reply]
Further, I absolutely disagree with the statement "_so in a way these teams are eating our own dogfood_" Going for WebView2 is anything but eating their own dog food.  It is a safe escape back to the HTML/JS. or Blazor, which are all out of scope of WASDK.  This is not eating your own dog food.  It is exactly the opposite.

**Comment by myokeeh** [comment]
> ### Q: Can your team work with the Windows team to enable WinUI3 apps to be easily selectable as Kiosk apps in Windows Settings? It's very easy to do that with my current UWP app. I can't move on to using WinUI3 because of this limitation.
> Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/9946#discussioncomment-10541229)
> 
> **A:** The WinAppSDK is a vehicle for us to ship features from across the company and we depend on each team's priority and schedule to deliver features like this. We are in communication with the team about this need, but they are currently unable to work on this. We will update you on GitHub as we know more.
> 

This blocker is 2+ years old at this point.

**Comment by rtanabe999** [comment]
> A: Microsoft.WindowsAppRuntime.Boostrap.dll must be located within the app, as it initializes dynamic dependencies for the app to find the rest of the runtime in the framework package directory. Placing it in the framework package is a catch-22.

I understand why Microsoft.WindowsAppRuntime.Bootstrap.dll cannot be included in the Windows App SDK runtime.
Now we would like to be able to statically link Microsoft.WindowsAppRuntime.Bootstrap.dll instead.
I am creating an app in C++ and the problem is that I have to distribute two files, an exe and a dll.

**Comment by Petrarca181** [comment]
Let me summarize this for you guys:

`[They are currently unable to work on this; We will update you; We don't have details for that at the moment; We don't have anything to share about plans for these right now; Unfortunately we're unable to prioritize; This isn't an area that we're focused on today; We can't prioritize this right now; We're hoping; Have not been able to prioritize that; We don't have plans; We love your enthusiasm!; This isn't currently in our plans but remains on our radar; We invite you to continue upvoting these Issues; ]`

If you have further questions pick a random line from the text above.

**Reply by whiskhub** [reply]
Still better than just silence.
It's clear the team has a lack of resource investment by Microsoft. Probably they will only get allocated with more resources once Windows 12 is coming with a new design...

**Reply by nikolayvpavlov** [reply]
No, they will not.  As I mentioned above, MS teams are simply not committed to WASDK.  Honestly, the new Outlook is a web atrocity, which should have never seen the light of day, instead MS are trying to shove it down our throats.  Even the new Photos app uses WebView2 to run its most sophisticated piece, the image editor.  How worse can it get?  Get a web-based Windows Explorer?

**Reply by Petrarca181** [reply]
Win 12 explorer will be on edge browser with bing AI search.

**Reply by 4OTbIPE** [reply]
> Win 12 explorer will be on edge browser with bing AI search.

All win12 UI will be on webview2 i bet)

**Reply by pjmlp** [reply]
@nikolayvpavlov Remember when folks from Project Reunion used to say WinUI was for wonderful engaging experiences, and everything else for classical? 

Why Photos uses Web,

> WebGL support enabling improved image rendering quality.


> Superior performance when sharing high quality images between the native and web layers using SharedBuffer.

So apparently DirectX and Win2D weren't good enough?!

**Reply by haidar47x** [reply]
Win 12 will be a PWA running on WebView, hidden behind a paywall.

**Comment by huoyaoyuan** [comment]
> **A:** We believe needing to set a XamlRoot is fully necessary for apps to handle, and it's also consistent with Win32 apps needing to specify a parent window. We will continue to work on improving error messages and code analyzers to help identify issues and fixes.

I don't agree with this. `MessageBox.Show` on both WinForms and WPF retrieves current active Windows as the parent. It's the real responsibility of a UI framework, not a dumb wrapper of Win32.
